api/src/app.ts

```diff
@@ -0,0 +1,193 @@
+  },
+  409: {
+    title: "Conflicting versions requested",
+    detail:
+      "The Accept header and the request path each requested different, conflicting versions. Please ensure that only one version is specified. You can set both the path and header to the same version, or you can request a version using only the path or only the header.",
+  },
+};
+
+/**
+ * Adds a `version` property to Express.Request, so we can attach
+ * that metadata as we resolve the version.
+ */
+declare global {
+  namespace Express {
+    interface Request {
+      version: number;
+    }
+  }
+}
+
+app.use(function (request, response, next) {
+  let [statusCode, version] = resolveVersions(request);
+  if (statusCode == 200) {
+    /**
+     * If the first segment includes the version,
+     * rewrite the path without the version.
+     */
+    if (request.url.match(/^\/v\d+/)) {
+      request.url = `/${request.url.split("/").slice(2).join("/")}`;
+    }
+    request.version = version;
+    next();
+  } else {
+    return response.status(statusCode).json({
+      errors: [
+        {
+          status: statusCode,
+          title: versionMessages[statusCode].title,
+          detail: versionMessages[statusCode].detail,
+        },
+      ],
+    });
+  }
+});
+
+app.get(
+  "/grants",
+  awaitErrorHandlerFactory(
+    async (request: Request, response: Response, next: NextFunction) => {
+      switch (request.version) {
+        case 1: {
+          let grants: Grant[] = await Grant.findAll({ limit: 10 });
+          return response.status(200).json({ data: grants });
+        }
+        default: {
+          return response.status(500).json({
+            error: {
+              title: "Internal Server Error: No version implemented",
+              detail: `Missing implementation for v${request.version}.`,
+            },
+          });
+        }
+      }
+    }
+  )
+);
+
+/**
+ * If no version is given, return 400 Bad Request.
+ * If conflicting versions are given, return 409 Conflict.
+ * If unavailable versions are given, return 404 Not Found.
+ * If a single valid version is given, pass through to the app.
+ */
+function resolveVersions(request: Request): number[] {
+  let headerVersion: number;
+  let pathVersion: number;
+
+  let versions: number[] = getVersions(request);
+  [headerVersion, pathVersion] = versions;
+
+  if (headerVersion == undefined && pathVersion == undefined) {
+    return [400, undefined];
+  } else if (
+    typeof headerVersion == "number" &&
+    typeof pathVersion == "number"
+  ) {
+    if (headerVersion == pathVersion) {
+      return [200, headerVersion];
+    } else {
+      return [409, undefined];
+    }
+  } else {
+    const version = headerVersion || pathVersion;
+    if (apiVersions.includes(version)) {
+      return [200, version];
+    } else {
+      return [404, undefined];
+    }
+  }
+}
+
+function getVersions(request: Request): number[] {
+  let headerVersion: number;
```

Comment:
**nitpick:** Can we document here what the return is in the case that a version is not specified?

Reply:
Done.